webapp/src/js/services/hydrate-messages.js

```diff
@@ -0,0 +1,67 @@
+var _ = require('underscore');
+
+angular.module('inboxServices').factory('HydrateMessages',
+  function(
+    $q,
+    DB,
+    LineageModelGenerator
+  ) {
+
+    'use strict';
+    'ngInject';
+
+    var hydrateMessage = function(doc, contactId, reportedDate) {
+      var contact = null, phone = null, message = null;
+      if(doc.kujua_message) {
+        var task = _.find(doc.tasks, function(task) {
+          var msg = task.messages[0];
+          if(msg.contact) {
+            return msg.contact._id === contactId;
+          }
+          return msg.to === contactId;
+        });
+        if(task && task.messages) {
+          message = task.messages[0].message;
+          contact = task.messages[0].contact;
+          phone = task.messages[0].to;
+        }
+      } else if (doc.sms_message) {
+        message = doc.sms_message.message;
+        contact = doc.contact;
+        phone = doc.from;
+      }
+
+      var reportPromise = LineageModelGenerator.reportSubject(contactId)
+        .catch(function(err) {
+          if(err.code !== 404) {
+            throw err;
+          }
+          return $q.resolve(null);
+        });
+
+      return reportPromise.then(function(report) {
+        var lineage = report && _.pluck(report.lineage, 'name');
+        return {
+          doc: doc,
+          id: doc._id,
+          key: contactId,
+          contact: report && report.doc.name,
+          lineage: lineage || [],
+          outgoing: doc.kujua_message ? true : false,
+          from: (contact && contact._id) || phone || doc._id,
+          date: reportedDate,
+          type: contact ? 'contact' : phone ? 'phone' : 'unknown',
```

Comment:
This would be easier to read without the nested ternaries. Break this out into a switch statement or if-else block outside of the return statement.